### DISCUSSION TITLE
Surface plot Z axis scale

### DISCUSSION THREAD

**Discussion by drod31** [bwoodsend/vtkplotlib#6] [Q&A]
I'm attempting to plot terrain data using a surface plot however the Z-axis scale seems to always distort my visual. For context, my latitude (y) region is between 50 - 70 and longitude (x) 140 -170. Typical Z axis values can range between 0 - 2000. Is there away to prevent the Z-axis from distorting the visual? 

Thanks, 

Daniel

**Comment by bwoodsend** [comment]
It shouldn't be distorting. In what way is it distorting? Stretching vertically? Shrinking vertically? Something else entirely?

> For context, my latitude (y) region is between 50 - 70 and longitude (x) 140 -170. Typical Z axis values can range between 0 - 2000.

These sound rather odd for terrain data. I would expect a terrain to be much wider than it is tall but yours is the opposite. It's like you've taken a very small snapshot of a very steep cliff? Is it possible that your longitudes and latitudes are in kilometers and your heights in meters? Or miles and feet?

If any of your code and/or data isn't considered confidential, it would be a lot easier if you could post it. Or failing that, maybe a screenshot of it distorting.

**Reply by drod31** [reply]
It's distorting in the z-axis (stretching). My dataset includes Alaska with 3e6< data points. 

Lats and Lons are in degrees and elevation is in meters. 

Some additional information: 

~1/3 of the data points are 0 elevation 

quick elevation stats: 
mean = ~300
median = ~85 

Could the sheer number of zeros in the elevation data be causing the distortion (stretching)?

**Reply by bwoodsend** [reply]
> Could the sheer number of zeros in the elevation data be causing the distortion (stretching)?

No but the mismatch in units will. You'll need to convert those degrees to meters to match your Z axis. Quoting [the sacred oracle](https://stackoverflow.com/a/1253545):

> * Longitude: 1 deg ~= 111.320*cos(latitude) km
> * Latitude: 1 deg ~= 110.574 km

So instead of using:
```python
vpl.surface(longitude, latitude, z)
```
Use:
```python
x = longitude * 111320 * np.cos(np.deg2rad(latitude))
y = latitude * 110574
vpl.surface(x, y, z)
```

**Reply by drod31** [reply]
It worked! Thanks! I didn't think of converting LLA to ECEF. 

Do you know how I can set the texture_map to use gist_earth or something similar? Currently the rendering appears as white.

**Reply by bwoodsend** [reply]
`gist_earth` is one dimensional so it's a colormap to vtkplotlib rather than a texturemap (although the jargon is very loose). You want the **cmap** parameter. And you'll need to generate some scalars although I'm guessing you're just wanting to color by height.
```python
# Color by elevation.
vpl.surface(x, y, z, cmap="gist_earth", scalars=z)
```
I would 2D describe texture maps as a work in progress but, given that I haven't had the stomach to finish them off for almost 2 years, I'm not sure it's fair to call them *in progress*.
